### DISCUSSION TITLE
Better support for Portrait displays

### DISCUSSION THREAD

**Discussion by gingerbeardman** [rxhanson/Rectangle#1833] [Old Issues]
I run a display that works in Landscape and Portrait and I switch depending on my workflow.

It would be great if the Rectangle functions that operate on "Halves" and "Thirds" of the screen could do so on the longest edge. That way I could organise in thirds horizontally when in landscape, and vertically when in portrait.

This feature would just be extra logic in the app, and would be transparent to most users.

Thanks for you consideration!

Currently, horizontal thirds on a portrait display do not work as expected, due to many apps minimum window width being larger than 1/3 the width of my display.

> If you implement a feature in a fork that you would like to get rolled into Rectangle, pull requests are still welcome but are not guaranteed to be accepted.

If you'd accept this feature request then I would be happy to try to implement it.

**Comment by rxhanson** [comment]
I'm actually not 100% sure what you're looking to change as I do not use displays in portrait mode, but I'd definitely lean toward accepting a pull request to fix the issues that you're experiencing. Thanks for checking up front.

Thirds should be properly handling portrait orientation, except it is admittedly strange behavior that the drag to snap edges are still the ones for landscape. See `getMouseHotSpot` in `SnappingManager` if that's what you are looking to fix there. If thirds aren't doing anything for portrait, then somethings wrong.

Halves don't handle portrait orientation. Being that I'm not a portrait mode user, my initial thought was that some users might still prefer halves done the way they are currently? Perhaps this is a feature that would be best done with a flag that could be set via Terminal.

**Comment by rxhanson** [comment]
Oh yeah, and on thirds not working due to minimum width > display width, could you post a screenshot? Of course, there's no way to make a window width smaller than it's minimum, but maybe there's something we can do?

**Comment by gingerbeardman** [comment]
My suggestion can be summarised as follows: portions of the screen should be calculated relative to longest edge, rather than being calculated assuming width is longest edge.

I appreciate there are also many smaller issues relating to landscape assumptions, but these are less of an issue for me at this point.

With the minimum width > 1/3 of display width, the window is aligned to the start of the third, and is cropped by the right edge of the display.

2 screens:

- leftmost 3rd
   ![Screen shot 2020-04-04 at 13 48 52](https://user-images.githubusercontent.com/49612/78451533-0b8d7c00-767e-11ea-9c6b-1ff212ad4b81.png)

- rightmost 3rd (notice part of window is obscured)
   ![Screen shot 2020-04-04 at 13 48 59](https://user-images.githubusercontent.com/49612/78451536-0e886c80-767e-11ea-9e05-feae5e03a856.png)

**Comment by rxhanson** [comment]
Ah ok, I think I see what you mean here. When you have portrait mode, you would expect the window on the right to be adjusted to stay on the screen but it is not doing that. Is that right?

[a comment by an automated account is not shown]

**Comment by rxhanson** [comment]
@finite-state-machine, thanks for the feedback. I'd like to zero in on the best way to approach this. I have a todo item for making more portrait icons for the menu bar menu (it's only thirds right now), and to make it so that they will reflect the display that the menu bar menu is open on - not just the focused display. I suspect that even with this, it would not be clear for everyone and there's a better place to communicate this. My thought is perhaps a line in the settings window that shows when you have a portrait display connected. I don't really want to add something like a warning popup to communicate this, if possible. Where in the UI would you have expected to see this communicated?

**Comment by gingerbeardman** [comment]
I ended up writing my own mini window manager using Hammerspoon.

**Reply by rxhanson** [reply]
Thanks for letting me know - sorry I never got to working on this. I'll be honest, I never could reproduce your original scenario with a window appearing off-screen, since Rectangle will default to moving windows on-screen if they don't fit and thirds are automatically set to top/center/bottom on portrait displays. I suspect there must have been an edge case that I was missing. I have a portrait display as part of my home setup now, so if you ever feel like revisiting Rectangle and rehashing this, let me know.

**Reply by gingerbeardman** [reply]
My old portrait display was 1200x1920 (non-retina) and very few apps windows min widths fit in 1200/3 or even 1200/2

My current portrait display is 1280x1920 (retina) which is not much difference but surprisingly many more windows fit
